tutorials-v5/time-evolution/024_v5_paper-mesolve.md

```diff
@@ -0,0 +1,593 @@
+plt.plot(tlist, driv_RWA_res.expect[0], "-.", label="mesolve (rwa)")
+plt.plot(tlist, np.real(results_corr_fit.expect[0]), "--", label=r"heomsolve")
+plt.plot(tlist, driv_br_res.expect[0], ":", linewidth=3, label="brmesolve")
+
+plt.xlabel(r"$t\, /\, \Delta^{-1}$")
+plt.ylabel(r"$\langle \sigma_z \rangle$")
+plt.legend()
+plt.show()
+```
+
+### Adiabatic Energy Switching
+
+To show where usage of these local-basis collapse operators can lead to errors, we look at a single qubit system.
+This time however, the energies of the qubit are adiabatically switched between positive and negative values.
+
+$$H = \dfrac{\Delta}{2} \sin{(\omega_d t)} \sigma_z.$$
+
+If the drive is slow enough, we expect the bath to respond to this change and transitions from higher to lower energy levels should be induced.
+These changes in system energy thus have to be taken into account so that `mesolve()` still produces a correct result.
+
+```python
+# Hamiltonian
+omega_d = 0.05 * Delta  # drive frequency
+A = Delta  # drive amplitude
+H_adi = [[A / 2.0 * sigmaz(), f]]
+
+# Bath parameters
+gamma = 0.05 * Delta / (2 * np.pi)
+
+# Simulation parameters
+T = 2 * np.pi / omega_d  # period length
+tlist = np.linspace(0, 2 * T, 400)
+```
+
+```python
+# Simple mesolve
+c_ops_me = [np.sqrt(gamma) * sigmam()]
+adi_me_res = mesolve(H_adi, psi0, tlist, c_ops=c_ops_me, e_ops=e_ops)
+```
+
+```python
+# HEOM
+HEOM_corr_fit = HEOMSolver(
+    QobjEvo(H_adi),
+    (exp_bath, sigmax()),
+    max_depth=max_depth,
+    options={"nsteps": 15000, "rtol": 1e-12, "atol": 1e-12},
+)
+adi_corr_fit_res = HEOM_corr_fit.run(psi0 * psi0.dag(), tlist, e_ops=e_ops)
+```
+
+```python
+# BRSolve non-flat power spectrum
+a_ops_non_flat = [[sigmax(), exp_bath]]
+brme_result = brmesolve(H_adi, psi0, tlist, a_ops=a_ops_non_flat, e_ops=e_ops)
+```
+
+```python
+# BRSolve
+a_ops = [[sigmax(), power_spectrum]]
+brme_result2 = brmesolve(H_adi, psi0, tlist, a_ops=a_ops, e_ops=e_ops)
+```
+
+```python
+plt.plot(tlist, adi_me_res.expect[0], "-", label="mesolve")
+plt.plot(tlist, np.real(adi_corr_fit_res.expect[0]), "--", label=r"heom")
+plt.plot(tlist, brme_result.expect[0], ":", linewidth=6, label="br non-flat")
+plt.plot(tlist, brme_result2.expect[0], ":", linewidth=6, label="br")
+
+plt.xlabel(r"$t\, /\, \Delta^{-1}$", fontsize=18)
+plt.ylabel(r"$\langle \sigma_z \rangle$", fontsize=18)
+plt.legend()
+plt.show()
+```
+
+## References
+
+\[1\] [QuTiP 5: The Quantum Toolbox in Python](https://arxiv.org/abs/2412.04705)
+
+
+## About
+
+```python
+about()
+```
+
+## Testing
+
+```python
+br_exp = br_res.expect[0]
+mg_exp = me_global_res.expect[0]
+np.testing.assert_allclose(br_exp, mg_exp, atol=0.2, rtol=0)
+
+d_exp = driv_res.expect[0]
+dr_exp = driv_RWA_res.expect[0]
+rc_exp = np.real(results_corr_fit.expect[0])
+db_exp = driv_br_res.expect[0]
+np.testing.assert_allclose(d_exp, dr_exp, atol=0.2, rtol=0)
+np.testing.assert_allclose(dr_exp, rc_exp, atol=0.2, rtol=0)
```

Comment:
btw, as written the tolerances are a bit too low, but reducing to atol=0.02 makes some tests fail. i think this is because the driven resonant example now uses  a larger gamma than in the original example in the paper, making the heom and master equations differ.   might be easier just to revert this gamma to the paper example and refit the bath with the energy switching example

Reply:
>  think this is because the driven resonant example now uses a larger gamma than in the original example in the paper

Just noticed, that the paper mentions `gamma = 0.005 Delta / 2pi` however, to generate the same plot as in the paper (and in accordance to the code files on overleaf) I needed to set `gamma = 0.05 Delta / 2pi`. So my guess is, that was a typo in the paper?

Reply:
Ah wait, this is actually only true for the driven example. So there are two different values for gamma at play here.